### PR TITLE
using bigdl-llm fused rope for llama

### PR DESCRIPTION
**Pull request by yangw1234**
## Description

using bigdl-llm fused rope for llama to reduce generation latency

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/llama.py [hunk 1]
```diff
@@ -103,9 +113,15 @@ def llama_attention_forward_4_31(
...
-    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
-    query_states, key_states = apply_rotary_pos_emb(query_states, key_states,
-                                                    cos, sin, position_ids, "llama")
+    # cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+    # query_states, key_states = apply_rotary_pos_emb(query_states, key_states,
+    #                                                 cos, sin, position_ids, "llama")
+    q_embed = torch.empty(query_states.shape, dtype=query_states.dtype, device=query_states.device)
+    k_embed = torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device)
+    
+    linear_q4_0.apply_rotary_embedding_half_qk(query_states, key_states, position_ids, q_embed, k_embed)
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/models/llama.py:122]: this only works on GPU, yes?

Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/models/llama.py:122]: I think cpu ipex has a similar function.

Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/models/llama.py:122]: > I think cpu ipex has a similar function.

We can add our GPU optimizations first

**Comment by yangw1234** [comment]
I think I am going to separate rms_norm, rope and other changes into different PRs with rms_norm being the first.

**Comment by yangw1234** [comment]
performance updated here https://github.com/analytics-zoo/nano/issues/606

@jason-dai would you mind reviewing again?

**Comment by yangw1234** [comment]
These optimizations does not work for training.

Should we check if it is in the training mode in every place? 
Or should we can assume user should not set `optimize_model=True` in training? 

@jason-dai

**Comment by yangw1234** [comment]
> These optimizations does not work for training.
> 
> Should we check if it is in the training mode in every place? Or should we can assume user should not set `optimize_model=True` in training?
> 
> @jason-dai

added checking for now

**Review by jason-dai — APPROVED** [review]
LGTM

**Pull request merged**